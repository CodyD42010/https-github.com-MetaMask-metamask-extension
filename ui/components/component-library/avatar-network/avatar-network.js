import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { AvatarBase } from '../avatar-base';
import Box from '../../ui/box/box';
import {
  Size,
  DISPLAY,
  AlignItems,
  JustifyContent,
  TextColor,
  BackgroundColor,
  BorderColor,
} from '../../../helpers/constants/design-system';
import { AVATAR_NETWORK_SIZES } from './avatar-network.constants';

export const AvatarNetwork = React.forwardRef(
  (
    {
      size = Size.MD,
      name,
      src,
      showHalo,
      color = TextColor.textDefault,
      backgroundColor = BackgroundColor.backgroundAlternative,
      borderColor = BorderColor.transparent,
      className,
      ...props
    },
    ref,
  ) => {
    const [showFallback, setShowFallback] = useState(false);

    useEffect(() => {
      setShowFallback(!src);
    }, [src]);

    const fallbackString = name && name[0] ? name[0] : '?';

    const handleOnError = () => {
      setShowFallback(true);
    };

    return (
      <AvatarBase
        ref={ref}
        size={size}
        display={DISPLAY.FLEX}
        alignItems={AlignItems.center}
        justifyContent={JustifyContent.center}
        className={classnames(
          'mm-avatar-network',
          showHalo && 'mm-avatar-network--with-halo',
          className,
        )}
        {...{ backgroundColor, borderColor, color, ...props }}
      >
        {showFallback ? (
          fallbackString
        ) : (
          <>
            {showHalo && (
              <img
                src={src}
                className={
                  showHalo ? 'mm-avatar-network__network-image--blurred' : ''
                }
                aria-hidden="true"
              />
            )}
            <img
              className={
                showHalo
                  ? 'mm-avatar-network__network-image--size-reduced'
                  : 'mm-avatar-network__network-image'
              }
              onError={handleOnError}
              src={src}
              alt={`${name} logo` || 'network logo'}
            />
          </>
        )}
      </AvatarBase>
    );
  },
);

AvatarNetwork.propTypes = {
  /**
   * The name accepts the string to render the first alphabet of the Avatar Name
   */
  name: PropTypes.string,
  /**
   * The src accepts the string of the image to be rendered
   */
  src: PropTypes.string,
  /**
   * The showHalo accepts a boolean prop to render the image with halo effect
   */
  showHalo: PropTypes.bool,
  /**
   * The size of the AvatarNetwork
   * Possible values could be Size.XS(16px), Size.SM(24px), Size.MD(32px), Size.LG(40px), Size.XL(48px)
   * Defaults to Size.MD
   */
  size: PropTypes.oneOf(Object.values(AVATAR_NETWORK_SIZES)),
  /**
   * The background color of the AvatarNetwork
   * Defaults to BackgroundColor.backgroundAlternative
   */
  backgroundColor: PropTypes.oneOf(Object.values(BackgroundColor)),
  /**
   * The background color of the AvatarNetwork
   * Defaults to BorderColor.borderDefault
   */
  borderColor: PropTypes.oneOf(Object.values(BorderColor)),
  /**
   * The color of the text inside the AvatarNetwork
   * Defaults to TextColor.textDefault
   */
  color: PropTypes.oneOf(Object.values(TextColor)),
  /**
   * Additional classNames to be added to the AvatarNetwork
   */
  className: PropTypes.string,
  /**
   * AvatarNetwork also accepts all Box props including but not limited to
   * className, as(change root element of HTML element) and margin props
   */
  ...Box.propTypes,
};

AvatarNetwork.displayName = 'AvatarNetwork';
