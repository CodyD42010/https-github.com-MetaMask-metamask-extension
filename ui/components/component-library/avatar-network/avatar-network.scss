.blur-halo-image {
  filter: blur(20px);
  transform: scale(1.25);
  border-radius: 50%;
  position: absolute;
  z-index: 1;
}
