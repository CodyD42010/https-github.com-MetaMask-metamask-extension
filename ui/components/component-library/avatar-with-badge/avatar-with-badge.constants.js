export const AVATAR_WITH_BADGE_POSTIONS = {
  TOP: 'top',
  BOTTOM: 'bottom',
};
