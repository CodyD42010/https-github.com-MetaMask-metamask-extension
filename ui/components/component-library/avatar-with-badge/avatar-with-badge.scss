.avatar-with-badge {
  position: relative;
  width: fit-content;

  &__token--position-top {
    position: absolute;
    top: -4px;
    right: -4px;
  }

  &__token--position-bottom {
    position: absolute;
    bottom: -4px;
    right: -4px;
  }
}
