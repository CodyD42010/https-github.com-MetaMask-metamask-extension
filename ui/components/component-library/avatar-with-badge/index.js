export { AvatarWithBadge } from './avatar-with-badge';
export { BADGE_POSITIONS } from './avatar-with-badge.constants';
