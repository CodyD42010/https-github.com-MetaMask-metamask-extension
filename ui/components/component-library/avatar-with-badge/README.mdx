import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { AvatarWithBadge } from './avatar-with-badge';

# AvatarWithBadge

The `AvatarWithBadge` is a wrapper component that adds badge display options to avatars.

<Canvas>
  <Story id="components-componentlibrary-avatarwithbadge--default-story" />
</Canvas>

## Props

The `AvatarWithBadge` accepts all props below as well as all [Box](/docs/components-ui-box--default-story#props) component props.

<ArgsTable of={AvatarWithBadge} />

### Badge Position

Use the `badgePosition` prop to set the position of the badge, it can have two values `Top` and `Bottom`

<Canvas>
  <Story id="components-componentlibrary-avatarwithbadge--badge-position" />
</Canvas>

```jsx
import { AvatarWithBadge } from '../ui/component-library';

<AvatarWithBadge
  badgePosition={AVATAR_WITH_BADGE_POSTIONS.BOTTOM}
  badge={
    <AvatarNetwork
      size={Size.XS}
      name="Arbitrum One"
      src="./images/arbitrum.svg"
    />
  }
>
  <AvatarAccount
    address="0x5CfE73b6021E818B776b421B1c4Db2474086a7e1"
    size={Size.MD}
    type={TYPES.JAZZICON}
  />
</AvatarWithBadge>

<AvatarWithBadge
  badgePosition={AVATAR_WITH_BADGE_POSTIONS.TOP}
  badge={
    <AvatarNetwork
      size={Size.XS}
      name="Arbitrum One"
      src="./images/arbitrum.svg"
    />
  }
>
  <AvatarAccount
    address="0x5CfE73b6021E818B776b421B1c4Db2474086a7e1"
    size={Size.MD}
    type={TYPES.JAZZICON}
  />
</AvatarWithBadge>
```

### Badge

Used to define the badge component to be rendered inside the `AvatarWithBadge`.

### Badge Props

The required props to be passed to the badge.

### Children

The children to be rendered inside the AvatarWithBadge. Generally used with the `AvatarAccount`
