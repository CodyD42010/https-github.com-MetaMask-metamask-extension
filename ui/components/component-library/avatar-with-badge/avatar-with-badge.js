import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import Box from '../../ui/box/box';
import { AVATAR_WITH_BADGE_POSTIONS } from './avatar-with-badge.constants';

export const AvatarWithBadge = ({
  children,
  badgePosition,
  className,
  badge,
  badgeWrapperProps,
  ...props
}) => {
  return (
    <Box className={classnames('mm-avatar-with-badge', className)} {...props}>
      {/* Generally the AvatarAccount */}
      {children}
      <Box
        className={
          badgePosition === 'top'
            ? 'mm-avatar-with-badge__badge-wrapper--position-top'
            : 'mm-avatar-with-badge__badge-wrapper--position-bottom'
        }
        {...badgeWrapperProps}
      >
        {/* Generally the AvatarNetwork at SIZES.XS */}
        {badge}
      </Box>
    </Box>
  );
};

AvatarWithBadge.propTypes = {
  /**
   * The position of the Badge
   * Possible values could be 'top', 'bottom',
   */
  badgePosition: PropTypes.oneOf(Object.values(AVATAR_WITH_BADGE_POSTIONS)),
  /**
   * The Badge Wrapper props of the component. All Box props can be used
   */
  badgeWrapperProps: PropTypes.shape(Box.PropTypes),
  /**
   * The children to be rendered inside the AvatarWithBadge
   */
  children: PropTypes.node,
  /**
   * The badge to be rendered inside the AvatarWithBadge
   */
  badge: PropTypes.object,
  /**
   * Add custom css class
   */
  className: PropTypes.string,
  /**
   * AvatarWithBadge accepts all the props from Box
   */
  ...Box.propTypes,
};
