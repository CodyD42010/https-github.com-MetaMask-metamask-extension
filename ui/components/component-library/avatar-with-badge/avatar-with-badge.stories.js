import React from 'react';
import { AvatarAccount } from '../avatar-account';
import { AVATAR_ACCOUNT_TYPES } from '../avatar-account/avatar-account.constants';
import { AvatarNetwork } from '../avatar-network';
import Box from '../../ui/box/box';
import {
  AlignItems,
  DISPLAY,
  Size,
} from '../../../helpers/constants/design-system';
import { AVATAR_WITH_BADGE_POSTIONS } from './avatar-with-badge.constants';
import README from './README.mdx';
import { AvatarWithBadge } from './avatar-with-badge';

export default {
  title: 'Components/ComponentLibrary/AvatarWithBadge',

  component: AvatarWithBadge,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    badgePosition: {
      options: Object.values(AVATAR_WITH_BADGE_POSTIONS),
      control: 'select',
    },
  },
  args: {
    badgePosition: AVATAR_WITH_BADGE_POSTIONS.top,
  },
};

export const DefaultStory = (args) => (
  <AvatarWithBadge
    badge={
      <AvatarNetwork
        size={Size.XS}
        name="Arbitrum One"
        src="./images/arbitrum.svg"
      />
    }
    {...args}
  >
    <AvatarAccount
      address="0x5CfE73b6021E818B776b421B1c4Db2474086a7e1"
      size={Size.MD}
      type={AVATAR_ACCOUNT_TYPES.JAZZICON}
    />
  </AvatarWithBadge>
);
DefaultStory.storyName = 'Default';

export const BadgePosition = () => (
  <Box display={DISPLAY.FLEX} alignItems={AlignItems.baseline} gap={1}>
    <AvatarWithBadge
      badgePosition={AVATAR_WITH_BADGE_POSTIONS.BOTTOM}
      badge={
        <AvatarNetwork
          size={Size.XS}
          name="Arbitrum One"
          src="./images/arbitrum.svg"
        />
      }
    >
      <AvatarAccount
        address="0x5CfE73b6021E818B776b421B1c4Db2474086a7e1"
        size={Size.MD}
        type={AVATAR_ACCOUNT_TYPES.JAZZICON}
      />
    </AvatarWithBadge>

    <AvatarWithBadge
      badgePosition={AVATAR_WITH_BADGE_POSTIONS.TOP}
      badge={
        <AvatarNetwork
          size={Size.XS}
          name="Arbitrum One"
          src="./images/arbitrum.svg"
        />
      }
    >
      <AvatarAccount
        address="0x5CfE73b6021E818B776b421B1c4Db2474086a7e1"
        size={Size.MD}
        type={AVATAR_ACCOUNT_TYPES.JAZZICON}
      />
    </AvatarWithBadge>
  </Box>
);
