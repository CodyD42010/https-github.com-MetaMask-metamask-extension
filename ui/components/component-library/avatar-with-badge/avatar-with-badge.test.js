/* eslint-disable jest/require-top-level-describe */
import { render } from '@testing-library/react';
import React from 'react';
import { AvatarNetwork } from '../avatar-network/avatar-network';
import { BorderColor } from '../../../helpers/constants/design-system';
import { AvatarWithBadge } from './avatar-with-badge';
import { AVATAR_WITH_BADGE_POSTIONS } from './avatar-with-badge.constants';

describe('AvatarWithBadge', () => {
  it('should render correctly', () => {
    const { getByTestId, container } = render(
      <AvatarWithBadge
        badgePosition={AVATAR_WITH_BADGE_POSTIONS.BOTTOM}
        data-testid="avatar-with-badge"
        badge={
          <AvatarNetwork
            name="Arbitrum One"
            src="./images/arbitrum.svg"
            data-testid="badge"
          />
        }
      />,
    );
    expect(getByTestId('avatar-with-badge')).toBeDefined();
    expect(getByTestId('badge')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it('should render badge network with bottom right position correctly', () => {
    const { container } = render(
      <AvatarWithBadge
        data-testid="avatar-with-badge"
        badgePosition={AVATAR_WITH_BADGE_POSTIONS.BOTTOM}
        badge={
          <AvatarNetwork
            name="Arbitrum One"
            src="./images/arbitrum.svg"
            data-testid="badge"
          />
        }
      />,
    );

    expect(
      container.getElementsByClassName(
        'mm-avatar-with-badge__badge-wrapper--position-bottom',
      ),
    ).toHaveLength(1);
  });

  it('should render badge network with top right position correctly', () => {
    const { container } = render(
      <AvatarWithBadge
        data-testid="avatar-with-badge"
        badgePosition={AVATAR_WITH_BADGE_POSTIONS.TOP}
        badge={
          <AvatarNetwork
            name="Arbitrum One"
            src="./images/arbitrum.svg"
            data-testid="badge"
          />
        }
      />,
    );

    expect(
      container.getElementsByClassName(
        'mm-avatar-with-badge__badge-wrapper--position-top',
      ),
    ).toHaveLength(1);
  });
  it('should render badge network with badgeWrapperProps', () => {
    const container = (
      <AvatarWithBadge
        data-testid="avatar-with-badge"
        badgePosition={AVATAR_WITH_BADGE_POSTIONS.TOP}
        badgeWrapperProps={{ borderColor: BorderColor.errorDefault }}
        badge={
          <AvatarNetwork
            name="Arbitrum One"
            src="./images/arbitrum.svg"
            data-testid="badge"
          />
        }
      />
    );
    expect(container.props.badgeWrapperProps.borderColor).toStrictEqual(
      'error-default',
    );
  });

  // className
  it('should render with custom className', () => {
    const { getByTestId } = render(
      <AvatarWithBadge
        data-testid="avatar-with-badge"
        className="test-class"
      />,
    );
    expect(getByTestId('avatar-with-badge')).toHaveClass('test-class');
  });
});
