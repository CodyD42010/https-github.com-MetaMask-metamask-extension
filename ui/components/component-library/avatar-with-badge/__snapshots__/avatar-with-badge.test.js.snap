// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarWithBadge should render correctly 1`] = `
<div>
  <div
    class="box mm-avatar-with-badge box--flex-direction-row"
    data-testid="avatar-with-badge"
  >
    <div
      class="box mm-avatar-with-badge__badge-wrapper--position-bottom box--flex-direction-row"
    >
      <div
        class="box mm-avatar-base mm-avatar-base--size-md mm-avatar-network box--display-flex box--flex-direction-row box--justify-content-center box--align-items-center box--color-text-default box--background-color-background-alternative box--border-color-transparent box--border-style-solid box--border-width-1"
        data-testid="badge"
      >
        <img
          alt="Arbitrum One logo"
          class="mm-avatar-network__network-image"
          src="./images/arbitrum.svg"
        />
      </div>
    </div>
  </div>
</div>
`;
